added gitattributes to disable autocrlf, addressing issue #1234

aceslowman opened the pull request:
Fixes/addresses #1234 (also #430)

Changes: 
 
I've added a .gitattributes file that disables git autocrlf on the repository. On Windows, git will convert LF to CRLF on checkout,  [which breaks findName() in the build script.](https://github.com/processing/p5.js-website/issues/430#issuecomment-474461235) Hopefully this will make this problem less likely to come back up for those on Windows.

takyano commented:
Should this be more focused and specify the file extensions that should be treated as text and binary?

    * text eol=lf

might treat everything as text including png, jpeg and other binary files.

aceslowman commented:
> Should this be more focused and specify the file extensions that should be treated as text and binary?
> 
> ```
> * text eol=lf
> ```
> 
> might treat everything as text including png, jpeg and other binary files.

That makes sense, I just pushed a new commit, targeting only .js files in data/examples, will it need to be more specific than that?

takyano commented:
What about any binary file formats like jpeg, gif, mov?

aceslowman commented:
> What about any binary file formats like jpeg, gif, mov?

Wouldn't binary files be unaffected by this? From what I understand, git should be able to detect whether or not a file should be treated as binary, at least how I understand [the discussion here](https://stackoverflow.com/questions/57030698/do-i-really-need-to-specify-all-binary-files-in-gitattributes).

takyano commented:
I'd want to be more explicit and not leave anything to chance.

I might just paranoid. Give it a try and see if anything gets broken.

aceslowman commented:
> I'd want to be more explicit and not leave anything to chance.
> 
> I might just paranoid. Give it a try and see if anything gets broken.

That's fair! Just updated the PR with a more explicit list of binary files.

takyano commented:
I did a scan of the p5.js-website git repo and got the following file extensions, separated by text and binary.

You could just include the binary files. Mainly adding the ogg, ogv and font files, otf and ttf

#	Cascading Style Sheet
css text eol=lf
#	Comma Separated Value
csv text eol=lf
#	Embedded JavaScript template file
ejs text eol=lf
#	ESLint configuration file
eslintrc text eol=lf
#	WebGL Fragment Shader file
frag text eol=lf

#	Some file that deals with translations
ftl text eol=lf

#	Handlebars template system file
hbs text eol=lf
#	Hyper Text Markup Language file
html text eol=lf
#	JavaScript source file
js text eol=lf
#	JavaScript Object Notation file
json text eol=lf
#	CSS map
map text eol=lf
#	Narkdown markup file
md text eol=lf
#	3D modei file
obj text eol=lf
#	PHP source file
php text eol=lf
#	Scalable Vector Graphics image file
svg text eol=lf
#	Tom's Obvious, Minimal Language file
toml text eol=lf
#	Text file
txt text eol=lf
#	WebGL Vertex Shader file
vert text eol=lf
#	Extensible Markup Language file
xml text eol=lf
#	YAML Ain't Markup Language file
yml text eol=lf




#	Adobe Illustrator file format
ai binary
#	Graphics Interchange Format file
gif binary
#	Windows Icon Image file
ico binary
#	Joint Photgraphic Experts Group image file
jpg binary
#	QuickTime media file container
mov binary
#	MP3 audio file
mp3 binary
#	MP4 media file container
mp4 binary
#	OGG media file container
ogg binary
#	OGG video file container
ogv binary
#	OpenType font file
otf binary
#	Portable Document Format file
pdf binary
#	Portable Network Grpahics image file
png binary
#	Truetype font file
ttf binary
#	Waveform Audio File Format audio file
wav binary
#	WebM media file
webm binary
#	ZIP archive file
zip binary

limzykenneth commented:
@aceslowman For a final check locally, you can try to normalize using the rules you just set and see if there are any unexpected changes: https://docs.github.com/en/get-started/getting-started-with-git/configuring-git-to-handle-line-endings#refreshing-a-repository-after-changing-line-endings

After that we can merge this. 😄

kjhollen reviewed (approved):
Looks good to me! Built this on my mac and verified sketches on examples and reference pages still run. I don't have access to a Windows machine, but will take Austin's word for it!

kjhollen commented:
@aceslowman can you try running the process that @limzykenneth suggested to check for errors? thanks! https://docs.github.com/en/get-started/getting-started-with-git/configuring-git-to-handle-line-endings#refreshing-a-repository-after-changing-line-endings

aceslowman commented:
@kjhollen Just went through the process! It all works on my end, I ran watch at the end and can navigate all areas of the site. '--renormalize' did leave me with some modified files though, should those be pushed as well?

kjhollen commented:
I think so—then they won't repeatedly show as modified for other contributors. Sounds like it's also a good opportunity to make sure that the remaining changed files are text files.

aceslowman commented:
@kjhollen The commit that I recently pushed to the branch showed these as the modified files, they all appear to be text files:

```
        modified:   offline-reference/extra/img/p5js-beta.svg
        modified:   offline-reference/extra/img/thick-asterisk-alone.svg
        modified:   src/assets/img/community/theartificial.svg
        modified:   src/assets/img/p5js-beta.svg
        modified:   src/assets/img/p5js.svg
        modified:   src/assets/img/thick-asterisk-alone-gray.svg
        modified:   src/assets/img/thick-asterisk-alone.svg
        modified:   src/assets/learn/basics/ellipse.svg
        modified:   src/assets/learn/basics/graph.svg
        modified:   src/assets/learn/basics/line.svg
        modified:   src/assets/learn/basics/rect.svg
        modified:   src/assets/learn/color/grayscale.svg
        modified:   src/assets/learn/curves/bezier/embed.html
        modified:   src/assets/learn/curves/bezier/sketch.js
        modified:   src/assets/learn/curves/curve_ex/embed.html
        modified:   src/assets/learn/curves/curve_ex/sketch.js
        modified:   src/assets/learn/interactivity/and.js
        modified:   src/assets/learn/interactivity/coded_keys.js
        modified:   src/assets/learn/interactivity/corners.js
        modified:   src/assets/learn/interactivity/cursor_1.js
        modified:   src/assets/learn/interactivity/cursor_2.js
        modified:   src/assets/learn/interactivity/cursor_3.js
        modified:   src/assets/learn/interactivity/draw_t.js
        modified:   src/assets/learn/interactivity/draw_t_2.js
        modified:   src/assets/learn/interactivity/event_1.js
        modified:   src/assets/learn/interactivity/event_2.js
        modified:   src/assets/learn/interactivity/invert_mouse.js        
        modified:   src/assets/learn/interactivity/key_1.js
        modified:   src/assets/learn/interactivity/key_2.js
        modified:   src/assets/learn/interactivity/key_code_1.js
        modified:   src/assets/learn/interactivity/key_code_2.js
        modified:   src/assets/learn/interactivity/key_var.js
        modified:   src/assets/learn/interactivity/mouse_2.js
        modified:   src/assets/learn/interactivity/mouse_3.js
        modified:   src/assets/learn/interactivity/mouse_boxes.js
        modified:   src/assets/learn/interactivity/mouse_pressed.js       
        modified:   src/assets/learn/interactivity/mouse_released.js      
        modified:   src/assets/learn/interactivity/moved_dragged.js       
        modified:   src/assets/learn/interactivity/pressed_1.js
        modified:   src/assets/learn/interactivity/pressed_2.js
        modified:   src/assets/learn/interactivity/pressed_3.js
        modified:   src/assets/learn/interactivity/prev_mouse.js
        modified:   src/assets/learn/interactivity/region_1.js
        modified:   src/assets/learn/interactivity/region_2.js
        modified:   src/assets/p5_featured/ChirsMill_AuroraII/AuroraII-p5front.js
        modified:   src/assets/p5_featured/Christine_Social_June/Bob.js   
        modified:   src/assets/p5_featured/Christine_Social_June/BobSystem.js
        modified:   src/assets/p5_featured/Christine_Social_June/equations.js
        modified:   src/assets/p5_featured/Christine_Social_June/index.html
        modified:   src/assets/p5_featured/Christine_Social_June/sketch.js        
        modified:   src/templates/pages/learn/curves.hbs
        modified:   src/templates/pages/learn/transformations.hbs
        modified:   src/templates/pages/reference/assets/blobs.csv
```

davepagurek commented on src/content/examples/en/12_Advanced_Canvas_Rendering/03_Shader_As_A_Texture_p5strands/code.js:
```diff
@@ -0,0 +1,81 @@
+}
+
+function shaderAsTextureCallback() {
+  const time = uniformFloat(() => millis() / 1000.0);
```
As in the earlier example, this doesn't actually need to be a uniform any more, inside of `pixelInputs` you should be able to write `const time = millis() / 1000)`.

davepagurek replied:
In p5.strands we also don't need the `.0` on numbers, `1000` will be fine.

davepagurek commented on src/content/examples/en/12_Advanced_Canvas_Rendering/03_Shader_As_A_Texture_p5strands/code.js:
```diff
@@ -0,0 +1,81 @@
+
+function shaderAsTextureCallback() {
+  const time = uniformFloat(() => millis() / 1000.0);
+  const size = uniformVec2(() => [width, height]);
```
I believe we can also just use `width` and `height` directly now without uniforms too!

davepagurek commented on src/content/examples/en/11_3D/07_Filter_Shader_p5strands/code.js:
```diff
@@ -0,0 +1,52 @@
+    return (coord - 0.5) / amount + 0.5;
+  }
+
+  getColor((inputs, canvasContent) => {
```
We've recently introduced a flatter p5.strands API that uses `filterColor.begin()` and `filterColor.end()` instead of `getColor` with a callback, since people have found the nested functions confusing. See some examples in these docs: https://beta.p5js.org/reference/p5/buildfiltershader/

Mind switching to that syntax in these examples? for most other `get*` hooks, the `get` part is just removed, e.g. `getWorldInputs` becomes just `worldInputs`.

davepagurek commented on src/content/examples/en/11_3D/08_Adjusting_Positions_With_A_Shader_p5strands/code.js:
```diff
@@ -0,0 +1,46 @@
+}
+
+function wiggleCallback() {
+  const time = uniformFloat(() => millis());
```
I believe this doesn't actually need to be a uniform any more, you can just use `millis()` directly inside of a strands shader 🙂

davepagurek commented on src/content/examples/en/11_3D/08_Adjusting_Positions_With_A_Shader_p5strands/code.js:
```diff
@@ -0,0 +1,46 @@
+function wiggleCallback() {
+  const time = uniformFloat(() => millis());
+
+  getObjectInputs((inputs) => {
```
Similar to before, can we use the flat API here?

davepagurek commented on src/content/examples/en/12_Advanced_Canvas_Rendering/03_Shader_As_A_Texture_p5strands/code.js:
```diff
@@ -0,0 +1,81 @@
+function shaderAsTextureCallback() {
+  const time = uniformFloat(() => millis() / 1000.0);
+  const size = uniformVec2(() => [width, height]);
+  // Y is flipped to match gl_FragCoord's bottom-up axis.
```
Partially a comment for @ksen0: do you think the main target audience of these examples is people looking to understand strands from a background of GLSL, or people just learning about strands who don't already know GLSL? For the latter, explaining how things translate to `gl_FragCoord` is introducing a point of reference that they aren't familiar with. If that's the goal, it might be better to just omit references to how `gl_FragCoord` worked in the old version of this?

davepagurek commented on src/content/examples/en/12_Advanced_Canvas_Rendering/03_Shader_As_A_Texture_p5strands/code.js:
```diff
@@ -0,0 +1,81 @@
+  const time = uniformFloat(() => millis() / 1000.0);
+  const size = uniformVec2(() => [width, height]);
+  // Y is flipped to match gl_FragCoord's bottom-up axis.
+  const mouse = uniformVec2(() => [mouseX, height - mouseY]);
```
Similarly, I think `mouseX` and `mouseY` should work without a uniform too! `const mouse = [mouseX, mouseY]` in a shader hook block should work.

davepagurek commented on src/content/examples/en/12_Advanced_Canvas_Rendering/03_Shader_As_A_Texture_p5strands/code.js:
```diff
@@ -0,0 +1,81 @@
+    return floor(distance(st, radius) * res) / scale;
+  }
+
+  getCameraInputs((inputs) => {
```
Can we switch these to the begin/end syntax?

davepagurek reviewed (commented):
Nice work converting these! This all works, so I just left some comments more about code style and using the most recent p5.strands features to simplify the code.

MissTipo commented:
> Nice work converting these! This all works, so I just left some comments more about code style and using the most recent p5.strands features to simplify the code.

Thanks Dave!  Making the changes. Will push updates soon.

ksen0 commented on src/content/examples/en/11_3D/08_Adjusting_Positions_With_A_Shader_p5strands/description.mdx:
```diff
@@ -0,0 +1,29 @@
+      - label: 2026 p5.strands translation
+        URL: https://github.com/processing/p5.js-website/tree/main/src/content/examples/en/11_3D/08_Adjusting_Positions_With_A_Shader_p5strands
+
+  - collectivelyAttributedSince: 2024
```
Hi Dorine! Please update this to 2026. So basically it's collectively attributed after it's published from you.

ksen0 replied:
Also in the other places where this appears

MissTipo replied:
Hi Kit! Quick heads up: changing collectivelyAttributedSince to 2026 needs a couple updates beyond the description.mdx files, otherwise Astro fails schema validation:
1. `src/content/examples/config.ts` — the schema needs to be updated to accept the new year value
2. `src/content/ui/*.yaml` files — a new attribution string needs to be added for the 2026 year

I've made both changes. Should I include them in this PR, or open a separate one?

ksen0 replied:
I'm okay with it being in this PR, I think it should be <20 lines? If more then maybe separate

ksen0 commented on src/content/examples/en/12_Advanced_Canvas_Rendering/02_Shader_As_A_Texture/description.mdx:
```diff
@@ -1,7 +1,7 @@
 featuredImage: "../../../images/featured/12_Advanced_Canvas_Rendering-02_Shader_As_A_Texture-thumbnail.png"
 featuredImageAlt: Two spheres broken up into a square grid with a gradient in each grid.
-title: Shader as a Texture
+title: Shader as a Texture(GLSL)
```
Minor: please add space before (

ksen0 commented:
These look wonderful thank you @MissTipo !

Some minor notes added.

I was also inspecting how it looks on the website, and I propose that we add the 3 strands sketches to featured; and remove 3 of the previous sketches except "orbitControl." This would bring more visibility to the strands examples. Other ideas on how to make this more navigable is welcome @davepagurek (also @nbogie maybe you have thoughts?)

<img width="1612" height="906" alt="image" src="https://github.com/user-attachments/assets/f988ccd3-eff3-45f4-826d-d9095f62edb7" />

<img width="1611" height="769" alt="image" src="https://github.com/user-attachments/assets/a490b317-f946-4da3-a327-071721741307" />

For implement featured solution: add/remove `featured: true` from the description mdx of the affected sketches.

Finally, all past sketches do have translations. Minor/can be done in a separate PR but everywhere were the description text is the same, the existing translations can be duplicated for the strands sketch, I think.

ksen0 reviewed (changes requested):
Minor

MissTipo commented:
> Finally, all past sketches do have translations. Minor/can be done in a separate PR but everywhere were the description text is the same, the existing translations can be duplicated for the strands sketch, I think.

Happy to handle the translation copying too! A quick clarification on titles, should I append (p5.strands) to the title in each new strands translated file, and correspondingly append (GLSL) to the title in each existing translated file as with the English description or should everything stay as is?

ksen0 commented:
Thanks @MissTipo ! Yes please update the titles to include the (GLSL) and (p5.strands). Thank you!

cc @Divyansh013 for visibility - I think for GLSL and p5.strands those are technical terms and can stay in English but please let me know if there are any other ideas you have?

davepagurek reviewed (commented):
The code updates look great, nice work @MissTipo!
